### PR TITLE
Add import/export VRML

### LINKED ISSUE #2438: VRML export
**Describe what feature you would like added.**

A clear and concise description of what you would like added.

If you are requesting that a feature from VTK be wrapped or exposed, please include links to the VTK documentation, example, or class definition.

Feel free to include pseudocode or screenshots of the requested outcome.

-----
Export the plotter to VRML format.

VTK feature
https://vtk.org/doc/nightly/html/classvtkVRMLExporter.html

### PR DESCRIPTION
**Pull request by tkoyama010**
### Overview

Add import/export VRML method to Plotter class

 

resolves #2438

Example

https://github.com/pyvista/pyvista/blob/2ee9c05fcba14cf0cceb22350d97d8fdee2c5a34/tests/example_files/Box.wrl#L1-L9

```python
>>> pl = pyvista.Plotter()
>>> pl.import_vrml("Box.wrl")
>>> pl.show()
```

import_vrml

### Details

- None

### CODE REVIEW

File: pyvista/plotting/plotting.py [hunk 1]
```diff
@@ -570,6 +570,41 @@ def export_gltf(self, filename, inline_data=True, rotate_scene=True, save_normal
...
+        """Export the current rendering scene as a VRML file.
+
+        See https://vtk.org/doc/nightly/html/classvtkVRMLExporter.html
+        for limitations regarding the exporter.
+
+        Parameters
+        ----------
+        filename : str
+            Filename to export the scene to.  A filename extension of
+            ``'wrl'`` will be added.
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:583]: Is the extension added even if it already exists? If so, we might want to add that as a note (or find a way of disabling it if we specify an extension).

File: pyvista/plotting/plotting.py [hunk 2]
```diff
@@ -570,6 +570,41 @@ def export_gltf(self, filename, inline_data=True, rotate_scene=True, save_normal
         for array in renamed_arrays:
             array.SetName('Normals')
 
+    def export_vrml(self, filename):
+        """Export the current rendering scene as a VRML file.
+
+        See https://vtk.org/doc/nightly/html/classvtkVRMLExporter.html
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:576]:
```suggestion
        See `vtk.VRMLExporter <https://vtk.org/doc/nightly/html/classvtkVRMLExporter.html>`_
```

File: pyvista/plotting/plotting.py [hunk 3]
```diff
@@ -570,6 +570,41 @@ def export_gltf(self, filename, inline_data=True, rotate_scene=True, save_normal
...
+        else:
+            filename = os.path.abspath(os.path.expanduser(filename))
+
+        # lazy import here to avoid importing unused modules
+        from vtkmodules.vtkIOExporter import vtkVRMLExporter
+
+        exporter = vtkVRMLExporter()
+        exporter.SetFileName(filename)
+        exporter.SetRenderWindow(self.ren_win)
+        return exporter.Write()
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:607]: Does this return anything?

File: pyvista/plotting/plotting.py (hunk 1 above)
Comment (tkoyama010) [reply pyvista/plotting/plotting.py:583]: I'm using the same style as the `export_vtkjs` comment, but it's not correct. Thanks for pointing that out.
```suggestion
            Filename to export the scene to.
```

File: pyvista/plotting/plotting.py (hunk 3 above)
Comment (tkoyama010) [reply pyvista/plotting/plotting.py:607]: Nothing is returned. Thank you.
```suggestion
        exporter.Write()
```

File: pyvista/plotting/plotting.py [hunk 4]
```diff
@@ -538,6 +570,42 @@ def export_gltf(self, filename, inline_data=True, rotate_scene=True, save_normal
...
+        Parameters
+        ----------
+        filename : str
+            Filename to export the scene to.  A filename extension of
+            ``'wrl'`` will be added.
+
+        Returns
+        -------
+        vtkOBJExporter
+            Object exporter.
```
Comment (tkoyama010) [inline pyvista/plotting/plotting.py:588]:
```suggestion
```

File: pyvista/examples/vrml.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        Filename of the VRML file.
+
+    Examples
+    --------
+    >>> import pyvista
+    >>> from pyvista import examples
+    >>> vrml_file = examples.vrml.download_sextant() # doctest:+SKIP
+    >>> pl = pyvista.Plotter() # doctest:+SKIP
+    >>> pl.import_vrml(vrml_file) # doctest:+SKIP
+    >>> pl.show() # doctest:+SKIP
```
Comment (akaszynski) [inline pyvista/examples/vrml.py:55]: I think we should allow doctests of these examples.

File: tests/plotting/test_plotting.py [hunk 6]
```diff
@@ -193,6 +193,32 @@ def test_export_gltf(tmpdir, sphere, airplane, hexbeam):
...
+def test_export_vrml(tmpdir, sphere, airplane, hexbeam):
+    filename = str(tmpdir.mkdir("tmpdir").join("tmp.wrl"))
+
+    pl = pyvista.Plotter()
+    pl.add_mesh(sphere, smooth_shading=True)
+    pl.export_vrml(filename)
+
+    pl_import = pyvista.Plotter()
+    pl_import.import_vrml(filename)
+    pl_import.show(before_close_callback=verify_cache_image)
```
Comment (akaszynski) [inline tests/plotting/test_plotting.py:216]: Good test!

File: pyvista/examples/__init__.py [hunk 7]
```diff
@@ -4,3 +4,4 @@
 from pyvista.examples.examples import *
 from . import gltf
 from . import hdr
+from . import vrml
```
Comment (akaszynski) [inline pyvista/examples/__init__.py:7]: We're not adding in the examples in our documentation at the moment. Let's add that in so they get rendered and we can see these new VRML examples in the documentation.

File: pyvista/examples/vrml.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
+"""VRML examples."""
+
+from .downloads import _retrieve_file
+
+VRML_SAMPLES_ROOT_URL = "https://raw.githubusercontent.com/lorensen/VTKExamples/master"
```
Comment (akaszynski) [inline pyvista/examples/vrml.py:5]: Agree with isolating the examples into a separate module. Our examples module is getting too large.

File: tests/plotting/test_plotting.py (hunk 6 above)
Comment (tkoyama010) [reply tests/plotting/test_plotting.py:216]: Thanks

File: pyvista/examples/__init__.py (hunk 7 above)
Comment (tkoyama010) [reply pyvista/examples/__init__.py:7]: That's a good idea. I'll add it in this PR.

File: pyvista/examples/vrml.py (hunk 5 above)
Comment (tkoyama010) [reply pyvista/examples/vrml.py:55]:
```suggestion
    >>> vrml_file = examples.vrml.download_sextant()
    >>> pl = pyvista.Plotter()
    >>> pl.import_vrml(vrml_file)
    >>> pl.show()
```

**Review by akaszynski — APPROVED** [review]
LGTM.

Recommending merge on 1 June 2022 unless further comments are given or issues raised.

**Pull request merged**